### PR TITLE
JWT better error handling

### PR DESCRIPTION
**Pull request by marcospri**
Current behavior of the JWT service is to fail silently and return an empty dictionary for invalid JWTs.

Raising `lms.ValidationError` instead which produces a meaningful error message.

## Testing 

- Simulate the three scenarios (not all at the same time):

```diff
--git a/lms/services/jwt.py b/lms/services/jwt.py
index 8beca8fb..a9f9e322 100644
--- a/lms/services/jwt.py
+++ b/lms/services/jwt.py
@@ -73,7 +73,7 @@ class JWTService:
         if not id_token:
             return {}
 
-        if not jwt.get_unverified_header(id_token).get("kid"):
+        if jwt.get_unverified_header(id_token).get("kid"):
             LOG.debug("Missing 'kid' value in JWT header")
             return {}
 
@@ -82,7 +82,7 @@ class JWTService:
 
         # Find the registration based on the token's claimed issuer & audience
         registration = self._registration_service.get(iss, aud)
-        if not registration:
+        if registration:
             LOG.debug("Unknown registration for lti_token. iss:%s aud:%s.", iss, aud)
             return {}
 
@@ -92,7 +92,7 @@ class JWTService:
             ).get_signing_key_from_jwt(id_token)
 
             return jwt.decode(
-                id_token, key=signing_key.key, audience=aud, algorithms=["RS256"]
+                "NOPE", key=signing_key.key, audience=aud, algorithms=["RS256"]
             )
         except PyJWTError as err:
             LOG.debug("Invalid JWT for: %s, %s. %s", iss, aud, str(err))
```

- On `main` this results in either a 

```You're not authorized to view this page```

in https://hypothesis.instructure.com/courses/319/assignments/3308

or a

```Sorry, but something went wrong. The issue has been reported and we'll try to fix it.```

on https://hypothesis.instructure.com/courses/319/assignments/3347 due to a ```KeyError: 'custom_canvas_course_id'``` on the logs. That was caused by the empty dictionary returned by the JWT service being interpreted as "This is an LTI1.1 launch".

- On this branch `jwt-error-handling` the same scenarios produce the following error messages in both assignment:

```
Hypothesis received an invalid request. There were problems with these request parameters:

jwt:
    Missing 'kid' value in JWT header
```

```
Hypothesis received an invalid request. There were problems with these request parameters:

jwt:
   Unknown registration for JWT. iss:https://canvas.instructure.com aud:93820000000000200.
```

```
Hypothesis received an invalid request. There were problems with these request parameters:

jwt:
    Invalid JWT for: https://canvas.instructure.com, 93820000000000200. Not enough segments
```

### CODE REVIEW

File: lms/services/jwt.py [hunk 1]
```diff
@@ -75,7 +76,9 @@ def decode_lti_token(self, id_token: str) -> dict:
 
         if not jwt.get_unverified_header(id_token).get("kid"):
             LOG.debug("Missing 'kid' value in JWT header")
```
Comment (seanh) [inline lms/services/jwt.py:78]: Double-check this but I think you might be able to remove these `LOG.debug()` calls now because raised exceptions get logged automatically

File: tests/unit/lms/security_test.py [hunk 2]
```diff
@@ -1,9 +1,8 @@
-from unittest.mock import PropertyMock, call, create_autospec, patch, sentinel
```
Comment (seanh) [inline tests/unit/lms/security_test.py:1]: Some unrelated changes here?

File: tests/unit/lms/services/jwt_test.py [hunk 3]
```diff
@@ -111,20 +112,23 @@ def test_decode_lti_token_with_empty_token(self, svc):
     def test_decode_lti_token_with_no_kid(self, svc):
         jwt_str = self.encode_jwt({"key": "value"}, headers={"key": "value"})
 
-        assert not svc.decode_lti_token(jwt_str)
+        with pytest.raises(ValidationError):
+            svc.decode_lti_token(jwt_str)
```
Comment (seanh) [inline tests/unit/lms/services/jwt_test.py:116]: You might also want to test that it raises the right error message? This can catch issues with either the code or the test set up (when a test isn't testing the situation that it thinks it is, but is actually triggering a different `ValidationError`). It frequently catches a mistake.

```python
with pytest.raises(ValidationError) as exc_info:
    svc.decode_lti_token(jwt_str)

assert exc_info.err.message == { ... }
```

**Review by seanh — APPROVED** [review]
:+1: Looks good, thanks to the previous work this is a very nice and simple change to make now.

It looks like there might be some changes from other PRs mixed in with this one?

One important question: are you sure these error messages aren't going to display any sensitive information from JWTs to the user?

File: lms/services/jwt.py (hunk 1 above)
Comment (marcospri) [reply lms/services/jwt.py:78]: The exception doesn't show up on the logs locally. I'll check in paper-trail when deployed

File: tests/unit/lms/security_test.py (hunk 2 above)
Comment (marcospri) [reply tests/unit/lms/security_test.py]: Not rebased, fixed now.

**Comment by marcospri** [comment]
> One important question: are you sure these error messages aren't going to display any sensitive information from JWTs to the user?

We are only logging the `aud` and `issuer` which are not that interesting. 

In any case nothing on the JWT is sensitive, it can be base64 decoded and all the information is there. It's only secure because we check it against the schools private keys.
